Source/Core/Core/Debugger/HLEFormatter.cpp

```diff
@@ -22,6 +29,232 @@ struct fmt::formatter<PowerPC::PairedSingle>
+        break;
+    }
+    // Only available in C++20
+    //
+    // return {type_start, last};
+    const std::size_t count = std::distance(type_start, last);
+    return {type_start, count};
+  }
+
+  template <typename T>
+  constexpr bool InArray(std::string_view s, T value) const
+  {
+    return s == value;
+  }
+
+  template <typename T, typename... Args>
+  constexpr bool InArray(std::string_view s, T value, Args... values) const
+  {
+    return InArray(s, value) || InArray(s, values...);
+  }
+
+  constexpr BaseType GetBaseType(std::string_view specifiers) const
+  {
+    // Requires C++20 as std::vector isn't constexpr
+    //
+    // auto in_array = [&](const std::vector<const char*>& values) {
+    //   return std::find(values.begin(), values.end(), specifiers) != values.end();
+    // };
+
+    auto in_array = [&](auto... values) { return InArray(specifiers, values...); };
+
+    if (in_array("c"))  // Char values
+      return BaseType::Char;
+    else if (in_array("s"))
+      return BaseType::String;
+    else if (in_array("hhd", "hhi"))  // Signed values
+      return BaseType::Int8;
+    else if (in_array("hd", "hi"))
+      return BaseType::Int16;
+    else if (in_array("d", "i"))
+      return BaseType::Int32;
+    else if (in_array("ld", "li"))
+      return BaseType::Int64;
+    else if (in_array("hhb", "hhB", "hho", "hhO", "hhu", "hhx", "hhX"))  // Unsigned values
+      return BaseType::Uint8;
+    else if (in_array("hb", "hB", "ho", "hO", "hu", "hx", "hX"))
+      return BaseType::Uint16;
+    else if (in_array("b", "B", "o", "O", "u", "x", "X"))
+      return BaseType::Uint32;
+    else if (in_array("lb", "lB", "lo", "lO", "lu", "lx", "lX"))
+      return BaseType::Uint64;
+    else if (in_array("a", "A", "e", "E", "f", "F", "g", "G"))  // Floating-point values
+      return BaseType::Float;
+    else if (in_array("la", "lA", "le", "lE", "lf", "lF", "lg", "lG"))
+      return BaseType::Double;
+    else if (in_array("p"))  // Pointer value
+      return BaseType::Pointer;
+    return BaseType::Default;
+  }
+
+  void TranslateBaseFormat()
+  {
+    // Convert the type specifiers to the ones supported by {fmt}
+    for (auto c = base_format.begin(); c != base_format.end();)
+    {
+      if (*c == 'u')
+      {
+        *c = 'd';
+      }
+      else if (*c == 'h' || *c == 'l')
+      {
+        c = base_format.erase(c);
+        continue;
+      }
+      ++c;
+    }
+  }
+
+  template <typename ParseContext>
+  constexpr auto parse(ParseContext& ctx)
+  {
+    const auto begin = ctx.begin();
+    const auto end = ctx.end();
+    if (begin == end || *begin == '}')
+    {
+      base_format = "{}";
+      return begin;
+    }
+
+    // Let {fmt} handle the invalid end iterator
+    const auto error = begin;
+
+    // Search the closing brace
+    const auto format_end = std::find(begin, end, '}');
+    if (format_end == end)
+      return error;
+    base_format = "{:" + std::string(begin, format_end + 1);
```

Comment:
I wish fmt  had more recent concatenation benchmarks.  I don't feel strongly about it but regardless, this could be:

`base_format = fmt::format("{:{}", std::string(begin, format_end + 1));`

Also, could you possibly put this down below?

```
if (base_type == BaseType::Pointer)
      base_format = "{:#010x}";
else
      base_format = fmt::format("{:{}", std::string(begin, format_end + 1));
```

Reply:
Good catch. Done.